### PR TITLE
Add test logs, trajectory plotting, and quintic spline interpolation

### PR DESCRIPTION
**Pull request by bhomberg**

### CODE REVIEW

File: exploration_scripts/trajectory_plotting [hunk 1]
```diff
@@ -0,0 +1,172 @@
+#!/usr/bin/env python
+from matplotlib import pyplot as plt
+import numpy as np
+import rosbag
+
+def get_waypoints(bag_name):
+    bag = rosbag.Bag(bag_name, 'r')
+    waypoints = None
+    for topic, msg, t in bag.read_messages():
+        print "Number of waypoints: ", len(msg.goal.trajectory.joint_trajectory.points)
```
Comment (jonbinney) [inline exploration_scripts/trajectory_plotting]: Use function form of `print()` instead of statement so this code will be compatible with python 3.

**Comment by bhomberg** [comment]
Addressed all comments, plus some additional cleanup.

File: exploration_scripts/trajectory_plotting [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+    positions = np.array([np.array(wp.positions) for wp in waypoints])
+    velocities = np.array([np.array(wp.velocities) for wp in waypoints])
+    accelerations = np.array([np.array(wp.accelerations) for wp in waypoints])
+    times = np.array([wp.time_from_start.to_sec() for wp in waypoints])
+    return (positions, velocities, accelerations, times)
+
+def calc_derivative_from_sequence(sequence, times):
+    derivatives = np.array([0.0, 0.0, 0.0, 0.0, 0.0, 0.0])
+    for i in range(len(sequence)-1):
```
Comment (jonbinney) [inline exploration_scripts/trajectory_plotting]: for iterating through trajectory points, use something like `point_i`. For iterating through joints within a trajectory point, use something like `joint_i`. This makes this code clearer whey you're indexing by i.

File: exploration_scripts/trajectory_plotting [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        if times[i+1] - times[i] == 0:
+            print "time delta is 0! skipping derivative"
+            derivative = np.array([0.0, 0.0, 0.0, 0.0, 0.0, 0.0])
+        else:
+            derivative = np.array([])
+            for j in range(len(sequence[i])):
+                derivative = np.append(derivative, (sequence[i+1][j] - sequence[i][j]) / (times[i+1] - times[i]))
+        derivatives = np.vstack((derivatives, derivative))
```
Comment (jonbinney) [inline exploration_scripts/trajectory_plotting:31]: Similar comment - probably easier to keep this as a list and append, then convert to an array at the end.

File: exploration_scripts/trajectory_plotting [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+    return derivatives
+
+def get_fig_and_axes():
+    fig, axes = plt.subplots(nrows=4, ncols=6, sharex=True, sharey=True)
```
Comment (jonbinney) [inline exploration_scripts/trajectory_plotting]: Seems like overkill to make this a function? don't need to change though.

File: exploration_scripts/trajectory_plotting [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    return fig, axes
+
+def plot_waypoints(positions, velocities, accelerations, jerks, times, axes, color, marker):
+    for i in range(6):
```
Comment (jonbinney) [inline exploration_scripts/trajectory_plotting]: `joint_i`

**Review by jonbinney — APPROVED** [review]

File: exploration_scripts/trajectory_plotting (hunk 1 above)
Comment (bhomberg) [reply exploration_scripts/trajectory_plotting]: Done

File: exploration_scripts/trajectory_plotting (hunk 2 above)
Comment (bhomberg) [reply exploration_scripts/trajectory_plotting]: Done

File: exploration_scripts/trajectory_plotting (hunk 5 above)
Comment (bhomberg) [reply exploration_scripts/trajectory_plotting]: Done

File: exploration_scripts/trajectory_plotting (hunk 4 above)
Comment (bhomberg) [reply exploration_scripts/trajectory_plotting]: Done

File: exploration_scripts/trajectory_plotting (hunk 3 above)
Comment (bhomberg) [reply exploration_scripts/trajectory_plotting:31]: Eh, I'll leave it as-is for now, it's already done and speed doesn't matter here.  I'll clean up if we use this for a while.

**Pull request merged**